app/schemas/tech.relaycorp.gateway.data.database.AppDatabase/1.json

```diff
@@ -119,12 +119,46 @@
           }
         ],
         "foreignKeys": []
+      },
+      {
+        "tableName": "ParcelAck",
+        "createSql": "CREATE TABLE IF NOT EXISTS `${TABLE_NAME}` (`recipientAddress` TEXT NOT NULL, `senderAddress` TEXT NOT NULL, `messageId` TEXT NOT NULL, PRIMARY KEY(`recipientAddress`, `senderAddress`, `messageId`))",
+        "fields": [
+          {
+            "fieldPath": "recipientAddress",
+            "columnName": "recipientAddress",
+            "affinity": "TEXT",
+            "notNull": true
+          },
+          {
+            "fieldPath": "senderAddress",
+            "columnName": "senderAddress",
```

Comment:
I think this should be `senderPrivateAddress` to be more specific -- it should never be set to a public address.

Reply:
Here the sender is coming from the outside, shouldn't it be public or private? Probably the recipient is the one that can only be private?

Reply:
> Here the sender is coming from the outside, shouldn't it be public or private?

The sender could indeed be a public or private endpoint. However, we can't tell from the outside whether the sender has a public address: We can only see its private address.

That's because, in a RAMF message, there's a metadata field for the recipient's private or public address, but no counterpart for the sender. Instead, the sender's private address is _inferred_ from the sender's certificate, which is contained in the digital signature -- Since private addresses are based on the public key in the certificate, we get it by calling `certificate.subjectPrivateAddress`.

The main reason for inferring the private address of the sender (instead of explicitly stating its private/public address in the metadata) is that the sender's public address (if any) is never used for routing or authorisation purposes.

> Probably the recipient is the one that can only be private?

Syntactically, the address could be either private or public. But because this is a private gateway, its public gateway should only be sending parcels bound for private endpoints.

However, we still have to send PCAs for invalid parcels so that the public gateway won't send them again. That means we may end up recording parcel collections for parcels bound for public endpoints -- even though those parcels are invalid and should be refused. I can think of two unlikely reasons why a public gateway would do that:

- A bug. The public gateway I implemented would refuse those invalid parcels -- But there could be a bug in the implementation.
- A rogue public gateway.